### ISSUE TITLE
PROGRAMFILESx86 record is messed up

### ISSUE THREAD

**Issue by maphew** [maphew/winfolders#1]
`print_all()` returns a path instead of description for PROGRAMFILESx86, and the wrong one at that. `fpath('PROGRAMFILESx86')` returns empty, as does `winsf.description()`

The rest of the records appear to be ok. So probably has something to do with it being the last row.

```
>>> winsf.print_all()
Name                 Path
DESKTOP              D:\Matt\Desktop
PROGRAMS             C:\Users\Matt\AppData\Roaming\Microsoft\Windows\Start Menu\Programs
...
PROGRAMFILESx86      C:\Users\Matt\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Administrative Tools
```

**Comment by maphew** [comment]
Nope, don't think it's a loop problem. The constants table I have (from [here](https://docs.microsoft.com/en-gb/windows/win32/api/shldisp/ne-shldisp-shellspecialfolderconstants#constants)) is out of step with what my Win10 computers are returning; 2 tested so far.

~~~
>>> shapp.namespace(47).self.path
'C:\\ProgramData\\Microsoft\\Windows\\Start Menu\\Programs\\Administrative Tools'

>>> shapp.namespace(48).self.path
'C:\\Users\\Matt\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Administrative Tools'

>>> shapp.namespace(49).self.path
'::{21EC2020-3AEA-1069-A2DD-08002B30309D}\\::{7007ACC7-3202-11D1-AAD2-00805FC1270E}'

>>> shapp.namespace(50).self.path
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'self'
~~~

So what we're actually getting is **FOLDERID_CommonAdminTools** and **FOLDERID_AdminTools** according to https://docs.microsoft.com/en-ca/windows/win32/shell/knownfolderid.

We _could_ just run through all integers from 0 to 50 or thereabouts and make semi-intelligent guesses about what we found. Unlikely to do that without any errors though. For confidence it looks like we'd need to figure out [SHGetKnownFolderIDList function](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderidlist) which suddenly makes this wee little weekend project a much bigger fish.

**Comment by maphew** [comment]
Ugh. might have to do it via ctypes. I don't understand that well enough to do it without cargo cult programming.
~~~
# win32com is missing SHGetKnownFolderIDList, so use ctypes.

_ole32 = ctypes.OleDLL('ole32')
_shell32 = ctypes.OleDLL('shell32')

_REFKNOWNFOLDERID = ctypes.c_char_p
_PPITEMIDLIST = ctypes.POINTER(ctypes.c_void_p)

_ole32.CoTaskMemFree.restype = None
_ole32.CoTaskMemFree.argtypes = (wintypes.LPVOID,)

_shell32.SHGetKnownFolderIDList.argtypes = (
    _REFKNOWNFOLDERID, # rfid
    wintypes.DWORD,    # dwFlags
    wintypes.HANDLE,   # hToken
    _PPITEMIDLIST)     # ppidl

def get_known_folder_id_list(folder_id, htoken=None):
    if isinstance(folder_id, pywintypes.IIDType):
        folder_id = bytes(folder_id)
    pidl = ctypes.c_void_p()
    try:
        _shell32.SHGetKnownFolderIDList(folder_id, 0, htoken,
            ctypes.byref(pidl))
        return shell.AddressAsPIDL(pidl.value)
    except WindowsError as e:
        if e.winerror & 0x80070000 == 0x80070000:
            # It's a WinAPI error, so re-raise it, letting Python
            # raise a specific exception such as FileNotFoundError.
            raise ctypes.WinError(e.winerror & 0x0000FFFF)
        raise
    finally:
        if pidl:
            _ole32.CoTaskMemFree(pidl)
~~~
Snippet courtesy of @eryk-sun https://stackoverflow.com/a/43635956/14420

**Comment by maphew** [comment]
it turns out brute forcing is useful, the new `hunter()` function cycles through all integer IDs given and returns `name` and `path` attributes. Unfortunately those appear to be the only 2 attributes available, so the goal of being locale agnostic can't be reached with this technique.

From what's returned here, it looks like the MSDN docs are out of date or wrong for Win10, ID:48 is definitely not PROGRAMFILESX86.

~~~
>>> sf = shapp.namespace(i).self

>>> sf._print_details_()
AxDispatch container <unknown>
Methods:
	 SetName
Props:
Get Props:
	Name = 0x0 - <win32com.client.build.MapEntry object at 0x0000019D146DC188>
	Path = 0x60020004 - <win32com.client.build.MapEntry object at 0x0000019D14186D08>
Put Props:
	Name = 0x0 - <win32com.client.build.MapEntry object at 0x0000019D1418A748>
~~~

~~~
>>> hunter(stop=70)
(0, 'Desktop', 'C:\\Users\\mattw\\Desktop')
(1, 'The Internet', '::{871C5380-42A0-1069-A2EA-08002B30309D}')
(2, 'Programs', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs')
(3, 'All Control Panel Items', '::{26EE0668-A00A-44D7-9371-BEB064C98683}\\0')
(4, 'Printers', '::{21EC2020-3AEA-1069-A2DD-08002B30309D}\\::{2227A280-3AEA-1069-A2DE-08002B30309D}')
(5, 'Documents', 'C:\\Users\\mattw\\Documents')
(6, 'Favorites', 'C:\\Users\\mattw\\Favorites')
(7, 'Startup', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup')
(8, 'Recent Items', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Recent')
(9, 'SendTo', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\SendTo')
(10, 'Recycle Bin', '::{645FF040-5081-101B-9F08-00AA002F954E}')
(11, 'Start Menu', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu')
(12, None, None)
(13, 'Music', 'C:\\Users\\mattw\\Music')
(14, 'Videos', 'C:\\Users\\mattw\\Videos')
(15, None, None)
(16, 'Desktop', 'C:\\Users\\mattw\\Desktop')
(17, 'This PC', '::{20D04FE0-3AEA-1069-A2D8-08002B30309D}')
(18, 'Network', '::{F02C1A0D-BE21-4350-88B0-7367FC96EF3C}')
(19, 'Network Shortcuts', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Network Shortcuts')
(20, 'Fonts', 'C:\\Windows\\Fonts')
(21, 'Templates', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Templates')
(22, 'Start Menu', 'C:\\ProgramData\\Microsoft\\Windows\\Start Menu')
(23, 'Programs', 'C:\\ProgramData\\Microsoft\\Windows\\Start Menu\\Programs')
(24, 'Startup', 'C:\\ProgramData\\Microsoft\\Windows\\Start Menu\\Programs\\StartUp')
(25, 'Public Desktop', 'C:\\Users\\Public\\Desktop')
(26, 'Roaming', 'C:\\Users\\mattw\\AppData\\Roaming')
(27, 'Printer Shortcuts', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Printer Shortcuts')
(28, 'Local', 'C:\\Users\\mattw\\AppData\\Local')
(29, 'Startup', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup')
(30, 'Startup', 'C:\\ProgramData\\Microsoft\\Windows\\Start Menu\\Programs\\StartUp')
(31, 'Favorites', 'C:\\Users\\mattw\\Favorites')
(32, 'INetCache', 'C:\\Users\\mattw\\AppData\\Local\\Microsoft\\Windows\\INetCache')
(33, 'INetCookies', 'C:\\Users\\mattw\\AppData\\Local\\Microsoft\\Windows\\INetCookies')
(34, 'History', 'C:\\Users\\mattw\\AppData\\Local\\Microsoft\\Windows\\History')
(35, 'ProgramData', 'C:\\ProgramData')
(36, 'Windows', 'C:\\Windows')
(37, 'System32', 'C:\\Windows\\System32')
(38, 'Program Files', 'C:\\Program Files')
(39, 'Pictures', 'C:\\Users\\mattw\\Pictures')
(40, 'mattw', 'C:\\Users\\mattw')
(41, 'SysWOW64', 'C:\\Windows\\SysWOW64')
(42, 'Program Files (x86)', 'C:\\Program Files (x86)')
(43, 'Common Files', 'C:\\Program Files\\Common Files')
(44, 'Common Files', 'C:\\Program Files (x86)\\Common Files')
(45, 'Templates', 'C:\\ProgramData\\Microsoft\\Windows\\Templates')
(46, 'Public Documents', 'C:\\Users\\Public\\Documents')
(47, 'Windows Administrative Tools', 'C:\\ProgramData\\Microsoft\\Windows\\Start Menu\\Programs\\Administrative Tools')
(48, 'Windows Administrative Tools', 'C:\\Users\\mattw\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Administrative Tools')
(49, 'Network Connections', '::{21EC2020-3AEA-1069-A2DD-08002B30309D}\\::{7007ACC7-3202-11D1-AAD2-00805FC1270E}')
(50, None, None)
(51, None, None)
(52, None, None)
(53, 'Public Music', 'C:\\Users\\Public\\Music')
(54, 'Public Pictures', 'C:\\Users\\Public\\Pictures')
(55, 'Public Videos', 'C:\\Users\\Public\\Videos')
(56, 'Resources', 'C:\\Windows\\Resources')
(57, None, None)
(58, 'OEM Links', 'C:\\ProgramData\\OEM Links')
(59, 'Temporary Burn Folder', 'C:\\Users\\mattw\\AppData\\Local\\Microsoft\\Windows\\Burn\\Burn')
(60, None, None)
(61, 'Network', '::{F02C1A0D-BE21-4350-88B0-7367FC96EF3C}')
(62, None, None)
(63, None, None)
(64, None, None)
(65, None, None)
(66, None, None)
(67, None, None)
(68, None, None)
(69, None, None)
(70, None, None)
~~~

**Comment by maphew** [comment]
OMG, it's been 5 years, and I've been here before: https://stackoverflow.com/questions/29888071/windows-special-and-known-folders-from-python-start-menu-downloads, 

and year later here: https://gist.github.com/mkropat/7550097

**Comment by maphew** [comment]
another complete ctypes implementation: https://stackoverflow.com/a/33181421/14420

**Comment by maphew** [comment]
Closing, the problem in OP is fixed in 96572f9, we had bad incoming data. 

However the whole approach is being changed and I will probably throughout the csv altogether. Opening a new issue or issues for that
